plugins/woocommerce/src/Templating/TemplatingEngine.php

```diff
@@ -0,0 +1,807 @@
+	 * "foo/bar" matches the "src/Templating/Templates/foo/bar.template" file.
+	 *
+	 * The complete path of the file will be passed to the woocommerce_template_file_path filter (null will be passed
+	 * if no file exists with the calculated path). This allows extensions to use their own templates.
+	 *
+	 * If the supplied $metadata is null, the result of the rendering will be returned directly as a string,
+	 * no new files will be created and no new entries will be added to the database.
+	 *
+	 * Otherwise, the result of the rendering will be stored as a file with a random name (customizable via
+	 * woocommerce_rendered_template_filename filter) in the directory whose name is the base location
+	 * (WordPress' uploads directory plus "woocommerce_rendered_templates", changeable via the
+	 * woocommerce_rendered_templates_directory filter) plus the expiration year and month formatted as 'yyyy-mm'.
+	 * The name of the file will be the return value of the function.
+	 *
+	 * $metadata must contain at least one of "expiration_date" (GMT, in the usual Y-m-d H:i:s format) or
+	 * "expiration_seconds" (a number that once added the current GMT date gives the expiration date).
+	 * The minimum resulting expiration date is "now + 60 seconds", this can be changed via the
+	 * woocommerce_rendered_template_minimum_expiration_seconds filter.
+	 *
+	 * $metadata must also contain an "is_public" key with the value `true` if the rendered files is intended
+	 * to be reachable via REST API.
+	 *
+	 * An entry for the file will be created in the wp_wc_rendered_templates table. Any metadata other than
+	 * "expiration_date", "expiration_seconds" and "is_public" will be stored as entries in the
+	 * wc_rendered_templates_meta table (these entries won't be used by the templating engine itself).
+	 *
+	 * @param string     $template_name The name of the template to render.
+	 * @param array      $variables Variables to be used for the rendering process, can be referenced in the template.
+	 * @param array|null $metadata Null to return the rendering result as a string, array to generate a rendered file with the corresponding database entry.
+	 * @return string The contents of the rendered template when rendering directly to a string, the name of the rendered file otherwise.
+	 * @throws Exception Filesystem or database error.
+	 * @throws InvalidArgumentException Template file not found, missing expiration date in $metadata, or wrong expiration date passed.
+	 * @throws OverflowException Template rendering depth of 255 levels reached, possible circular reference when rendering secondary templates.
+	 */
+	public function render_template( string $template_name, array $variables, ?array $metadata = null ): ?string {
+		global $wpdb;
+		global $wp_filesystem;
+
+		$render_to_file = ! is_null( $metadata );
+		$render_ok      = false;
+
+		// phpcs:disable WordPress.WP.AlternativeFunctions
```

Comment:
I think you might be able to do some, if not all, of the file operations in this method using `WP_Filesystem`, which is really just a wrapper around a lot of the same functions. The advantage is that is does some error handling for you, and handles some gotchas related to the quirks of varying file systems on different servers.

You just need to start by making sure the file system global [is initialized](https://github.com/woocommerce/woocommerce/pull/41114/files#diff-2dfe61449b3cb0e7224c2ffa46a023aebbdc4fa1d921bd29dfdf1ea424d74df9R62-R65), and then you can call its methods. Here are the ones that I think would pertain to `render_template`:

* [mkdir](https://github.com/WordPress/WordPress/blob/9c8d8c260a5fb14f5562c28bc1ca85c40494c443/wp-admin/includes/class-wp-filesystem-direct.php#L546)
* [delete](https://github.com/WordPress/WordPress/blob/9c8d8c260a5fb14f5562c28bc1ca85c40494c443/wp-admin/includes/class-wp-filesystem-direct.php#L373)
* [put_contents](https://github.com/WordPress/WordPress/blob/9c8d8c260a5fb14f5562c28bc1ca85c40494c443/wp-admin/includes/class-wp-filesystem-direct.php#L65)

Reply:
I see that `WP_Filesystem::mkdir` doesn't have the `$recursive` argument, which is a problem since I need to create a directory like `base/expiration`. Also I see that `WP_Filesystem::mkdir` is just resorting to the PHP built-in `mkdir` after all.

Reply:
As for `put_contents`, I can't use it. I need to open the file and use the file handle in the callback that is passed to `ob_start`.

Reply:
Well, instead of giving the output buffer a callback, you could just do something like this:

```
ob_start();
// output stuff
$content = ob_get_clean();
$wp_filesystem->put_contents( $file_path, $content );
```

Reply:
We don't know how big the rendered file will be (or put another way: we _can_ figure it out by looking the file size of the template, but it could be arbitrarily large) so it's better to process the output in chunks (letting the `ob_*` engine handle buffering by itself). That's also why in `TemplatingRestController::handle_parse_request` I'm sending the rendered file in chunks of 1KByte.